src/test/java/io/strimzi/kafka/bridge/mqtt/MqttBridgetIT.java

```diff
@@ -0,0 +1,252 @@
+    private static MqttServer mqttBridge;
+    private static Thread prepareServerThread;
+    private static StrimziKafkaContainer kafkaContainer;
+    private String mqttClientId;
+
+
+    /**
+     * Start the kafka Cluster
+     * Start the MQTT bridge before all tests
+     */
+    @BeforeAll
+    public static void beforeAll() throws InterruptedException {
+        final Map<String, Object> config = new HashMap<>();
+        String kafkaBootstrapServers = "localhost:9092";
+        try {
+            kafkaContainer = new StrimziKafkaContainer();
+            kafkaContainer.waitForRunning();
+            kafkaContainer.start();
+            kafkaBootstrapServers = kafkaContainer.getBootstrapServers();
+        } catch (Exception e) {
+            logger.error("Exception occurred", e);
+        }
+
+        // prepare the configuration for the bridge
+        config.put(KafkaConfig.KAFKA_CONFIG_PREFIX + ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaBootstrapServers);
+        config.put(MqttConfig.MQTT_HOST, MQTT_SERVER_HOST);
+        config.put(MqttConfig.MQTT_PORT, MQTT_SERVER_PORT);
+        config.put(BridgeConfig.BRIDGE_ID, "my-bridge");
+
+        // instantiate the kafka consumer client
+        kafkaConsumerClient = new KafkaConsumer<>(
+                ImmutableMap.of(
+                        ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaBootstrapServers,
+                        ConsumerConfig.GROUP_ID_CONFIG, "gid-" + UUID.randomUUID(),
+                        ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, OffsetResetStrategy.EARLIEST.name().toLowerCase(Locale.ROOT)
+                ),
+                new StringDeserializer(),
+                new StringDeserializer()
+        );
+
+        // consumer client subscribes to the kafka topic
+        kafkaConsumerClient.subscribe(Collections.singletonList(KAFKA_TOPIC));
+
+        // start the bridge server in a separate thread
+        prepareServerThread = new Thread(() -> {
+            try {
+                EventLoopGroup bossGroup = new NioEventLoopGroup();
+                EventLoopGroup workerGroup = new NioEventLoopGroup();
+
+                BridgeConfig bridgeConfig = BridgeConfig.fromMap(config);
+
+                String mappingRulesPath = Objects.requireNonNull(MqttBridgetIT.class.getClassLoader().getResource("mapping-rules-regex.json")).getPath();
+                MappingRulesLoader.getInstance().init(mappingRulesPath);
+
+                mqttBridge = new MqttServer(bridgeConfig, bossGroup, workerGroup, ChannelOption.SO_KEEPALIVE);
+                mqttBridge.start();
+            } catch (InterruptedException interruptedException) {
+                // in this case, we are interrupting the thread, so we don't need to do anything
+                logger.info("Shutting down the server");
+            } catch (Exception e) {
+                logger.error("Exception occurred", e);
+            }
+        });
+        prepareServerThread.start();
+        Thread.sleep(5000);
```

Comment:
What's the reason for this sleep? We don't like sleep :-)
If you have a sync problem waiting for the MQTT server being started we have to deal with it better.
Maybe it would mean having the bridge `start` returning a `CompletableFuture` and wait for it?

Reply:
I tried with CountDownLatch before, as I said in the previous comment. Maybe I was not using it appropriately. Let me try it here.
Yeah, we do not want to sleep.

If this sleep is not good, then sleep in @AfterAll should be removed too.

Reply:
@ppatierno I do not think using Latch fixes the issue.
Maybe I am missing something.

```java
....inside the thread block
// Do some task here
mqttServer.start();

// Signal that the task is completed
latch.countDown();
```
```
The latch will never count down because `.start()` blocks the thread. Of course, there is no sense to count down the latch before starting the server.

Reply:
As I read the `start` method ...

```java
ChannelFuture channelFuture = this.serverBootstrap.bind(host, port).sync();
channelFuture.channel().closeFuture().sync();
```

the first `sync` on the binding is actually waiting synchronously for the binding to be done. So you are actually making an async operation by Netty (the binding) to be sync.
Then next sync actually block the thread where you start the bridge (this is why in the tests you needed a new thread).
It means that in your tests the `start` is a blocking call and you will never be able to use a `CountDownLatch.countDown` right after the start, because it will never executed until the bridge is closed.
I think this is the reason why it didn't work.

So at this point I think the `start` method should be slightly changed. You should waiting for the bind and then completing a future returned by the `start` method itself to unlock the caller.
Right after that you can sync on the closeFuture and actually blocking the bridge thread.

Reply:
thanks sensei